### PR TITLE
Add a model fo WebVision 2018

### LINKED ISSUE #2774: add WebVision2018 code and models
add WebVision2018 code and models

### PR DESCRIPTION
**Pull request by yuanpengcheng**
resolve https://github.com/PaddlePaddle/models/issues/2774  add webvision models and code

### CODE REVIEW

File: PaddleCV/Research/webvision2018/infer.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+    image = fluid.layers.data(name='image', shape=image_shape, dtype='float32')
+
+    # model definition
+    model = models.__dict__[model_name]()
+    if model_name == "GoogleNet":
+        out, _, _ = model.net(input=image, class_dim=class_dim)
+    else:
+        out = model.net(input=image, class_dim=class_dim)
+        #out = fluid.layers.softmax(out)
```
Comment (jerrywgz) [inline PaddleCV/Research/webvision2018/infer.py]: remove useless code

File: PaddleCV/Research/webvision2018/infer.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+    test_program = fluid.default_main_program().clone(for_test=True)
+
+    fetch_list = [out.name]
+    if with_memory_optimization and not save_inference:
+        fluid.memory_optimize(
```
Comment (jerrywgz) [inline PaddleCV/Research/webvision2018/infer.py]: memory_optimize is deprecated and there are several methods in https://www.paddlepaddle.org.cn/documentation/docs/zh/1.5/advanced_usage/best_practice/memory_optimize.html

File: PaddleCV/Research/webvision2018/utils/learning_rate.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+            with switch.case(epoch < warmup_epoch):
+                decayed_lr = learning_rate * (global_step / (step_each_epoch * warmup_epoch))
+                fluid.layers.tensor.assign(input=decayed_lr, output=lr)
+            with switch.default():
+                decayed_lr = learning_rate * \
+                    (ops.cos((global_step - warmup_epoch * step_each_epoch) * (math.pi / (epochs * step_each_epoch))) + 1)/2
+                fluid.layers.tensor.assign(input=decayed_lr, output=lr)
+    return lr
+
+def lr_warmup(learning_rate, warmup_steps, start_lr, end_lr):
```
Comment (jerrywgz) [inline PaddleCV/Research/webvision2018/utils/learning_rate.py]: Now linear warmup strategy can be used here directly https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/fluid/layers/learning_rate_scheduler.py#L441

File: PaddleCV/Research/webvision2018/infer.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
+#copyright (c) 2019 PaddlePaddle Authors. All Rights Reserve.
```
Comment (jerrywgz) [inline PaddleCV/Research/webvision2018/infer.py]: We only release inference code and model?

File: PaddleCV/Research/webvision2018/infer.py (hunk 1 above)
Comment (yuanpengcheng) [reply PaddleCV/Research/webvision2018/infer.py]: already removed the useless code

File: PaddleCV/Research/webvision2018/infer.py (hunk 2 above)
Comment (yuanpengcheng) [reply PaddleCV/Research/webvision2018/infer.py]: I have remove these lines.

File: PaddleCV/Research/webvision2018/infer.py (hunk 4 above)
Comment (yuanpengcheng) [reply PaddleCV/Research/webvision2018/infer.py]: yes

File: PaddleCV/Research/webvision2018/utils/learning_rate.py (hunk 3 above)
Comment (yuanpengcheng) [reply PaddleCV/Research/webvision2018/utils/learning_rate.py]: have already remove the codes

File: PaddleCV/Research/webvision2018/utils/fp16_utils.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
+#copyright (c) 2019 PaddlePaddle Authors. All Rights Reserve.
```
Comment (qingqing01) [inline PaddleCV/Research/webvision2018/utils/fp16_utils.py]: 这个模型使用fp16训练了吗？

File: PaddleCV/Research/webvision2018/utils/utility.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
+"""Contains common utility functions."""
+#  Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserve.
```
Comment (qingqing01) [inline PaddleCV/Research/webvision2018/utils/utility.py]: Copyright一些是2018，一些是2019

File: PaddleCV/Research/webvision2018/reader_cv2.py [hunk 7]
```diff
@@ -0,0 +1,312 @@
...
+
+random.seed(0)
+np.random.seed(0)
+
+DATA_DIM = 224
+
+THREAD = 8
+BUF_SIZE = 2048
+
+DATA_DIR = './data/ILSVRC2012'
```
Comment (qingqing01) [inline PaddleCV/Research/webvision2018/reader_cv2.py]: 数据路径还是ILSVRC2012？ 数据路径最好参数里可配置。其他全局参数同。

File: PaddleCV/Research/webvision2018/utils/fp16_utils.py (hunk 5 above)
Comment (yuanpengcheng) [reply PaddleCV/Research/webvision2018/utils/fp16_utils.py]: 已删除，谢谢~

File: PaddleCV/Research/webvision2018/utils/utility.py (hunk 6 above)
Comment (yuanpengcheng) [reply PaddleCV/Research/webvision2018/utils/utility.py]: 已修改，谢谢

File: PaddleCV/Research/webvision2018/reader_cv2.py (hunk 7 above)
Comment (yuanpengcheng) [reply PaddleCV/Research/webvision2018/reader_cv2.py]: 已修改，已改成settings中参数配置

File: PaddleCV/Research/webvision2018/reader_cv2.py [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+import cv2
+import io
+
+import paddle
+import paddle.fluid as fluid
+
+random.seed(0)
+np.random.seed(0)
+
+DATA_DIM = 224
```
Comment (qingqing01) [inline PaddleCV/Research/webvision2018/reader_cv2.py]: 这个数下面用到了吗？

File: PaddleCV/Research/webvision2018/reader_cv2.py [hunk 9]
```diff
@@ -0,0 +1,299 @@
+#copyright (c) 2019 PaddlePaddle Authors. All Rights Reserve.
```
Comment (qingqing01) [inline PaddleCV/Research/webvision2018/reader_cv2.py]: 既然只有一个reader，为啥加 _cv2的后缀？

File: PaddleCV/Research/webvision2018/reader_cv2.py [hunk 10]
```diff
@@ -0,0 +1,299 @@
...
+        settings=settings,
+        mode=mode,
+        color_jitter=color_jitter,
+        rotate=rotate)
+
+    return paddle.reader.xmap_readers(
+        mapper, data_reader, THREAD, BUF_SIZE, order=False)
+
+
+def train(settings, batch_size, data_dir=None, shuffle_seed=0):
```
Comment (qingqing01) [inline PaddleCV/Research/webvision2018/reader_cv2.py]: 没有训练的流程的情况下还需要train吗？

**Review by qingqing01 — COMMENTED** [review]
不提供eval流程吗？

**Comment by yuanpengcheng** [comment]
> 不提供eval流程吗？

已补充eval代码，且测试结果与公布结果一致

File: PaddleCV/Research/webvision2018/reader_cv2.py (hunk 8 above)
Comment (yuanpengcheng) [reply PaddleCV/Research/webvision2018/reader_cv2.py]: 已删除，谢谢

File: PaddleCV/Research/webvision2018/reader_cv2.py (hunk 9 above)
Comment (yuanpengcheng) [reply PaddleCV/Research/webvision2018/reader_cv2.py]: 已修改，谢谢！

File: PaddleCV/Research/webvision2018/reader_cv2.py (hunk 10 above)
Comment (yuanpengcheng) [reply PaddleCV/Research/webvision2018/reader_cv2.py]: 已删除，谢谢

**Review by qingqing01 — APPROVED** [review]
